arbor/include/arbor/swcio.hpp

```diff
@@ -57,37 +71,59 @@ class swc_record {
             parent_id == other.parent_id;
     }
 
-    friend bool operator!=(const swc_record& lhs, const swc_record& rhs) {
-        return !(lhs == rhs);
+    bool operator!=(const swc_record& other) const {
+        return !(*this == other);
     }
 
-    friend std::ostream& operator<<(std::ostream& os, const swc_record& record);
-
-    coord_type diameter() const {
-        return 2*r;
-    }
-
-    arb::point<coord_type> coord() const {
-        return arb::point<coord_type>(x, y, z);
-    }
-
-    // validity checks
-    bool is_consistent() const;
-    void assert_consistent() const; // throw swc_error if inconsistent.
+    friend std::ostream& operator<<(std::ostream&, const swc_record&);
+    friend std::istream& operator>>(std::istream&, swc_record&);
 };
 
+struct swc_data {
+    std::string metadata;
+    std::vector<swc_record> records;
+};
 
-// Parse one record, skipping comments and blank lines.
-std::istream& operator>>(std::istream& is, swc_record& record);
-
-// Parse and canonicalize an EOF-terminated sequence of records.
-// Throw on parsing failure.
-std::vector<swc_record> parse_swc_file(std::istream& is);
+// Read SWC records from stream, collecting any initial metadata represented
+// in comments (stripping initial '#' and subsequent whitespace).
+// Stops at EOF or after reading the first line that does not parse as SWC.
+//
+// Note that 'one-point soma' SWC files are explicitly not supported.
+//
+// In `relaxed` mode, it will check that:
+//     * There are no duplicate record ids.
+//     * All record ids are positive.
+//     * There are no records whose parent id is not less than the record id.
+//     * Only one record has parent id -1; all other parent ids correspond to records.
+//     * There are at least two records.
+//
+// In `strict` mode, it will additionally check:
```

Comment:
This is too course grained.
E.g. Allen model insist that records on the same unbranched section should be contiguously numbers, and that the soma must be a "sphere/single sample", which can't be specified with this flag.

Reply:
It's definitely not meant to cover all cases: the intention of strict is to firstly abide by standard SWC conventions, and secondly prohibit spherical somata.

Relaxed just enforces, essentially, the bare minimum; Allen-specific checks would be imposed by whatever is dealing with their particular brand of SWC.

Reply:
Then why shouldn't all these checks be enforced in the different readers? Unfortunately there is no SWC spec that mandates the rules that are being enforced here, and in practice we see that tools like NEURON are far more relaxed. So, despite these being reasonable rules, they don't appear to be 'standard', and in order to make SWC a useful format folk will nearly never use these strict defaults.

Reply:
I don't think this is true: the 'relaxed' mode is, as far as I can tell, much more relaxed than any extant SWC description, and what's more, I think it corresponds to a minimal reasonable set of requirements, as far as sensible parsing is concerned.

The strict mode adds exactly two requirements: contiguous record numbering, which again I think is common to every SWC spec, and a prohibition of single-point somata. The latter is a necessary requirement for sensible interpretation by Arbor. That extant SWC files _do_ use single point somata is a problem we can't fix: we can only provide ad hoc interpretations as required at best.

In terms of parsing SWC and generating an unambiguous morphology, 'relaxed' + no single point soma is the bare minimum. I can see that would could relax the contiguous numbering requirement, but I don't think this is the problem, right? I included that in 'strict' only because of its commonality in SWC format descriptions, and because in my own view, when reading a complete SWC file (as opposed to a fragment), non-contiguous data is likely to indicate bad data.

tl;dr: 'relaxed' is really relaxed and should accept any extant SWC; 'strict' is slightly more strict than necessary for unambiguous interpretation, but that extra strictness I don't think precludes any extant SWC.

Reply:
I think that both of these tests are valuable, but I still feel that bundling them together isn't great for the following reason.

If contiguous numbering is common enough to be a kind of standard requirement, but single point somas are commonly accepted, it doesn't make sense to bundle those two requirements. A parser of the records that handles spherical somas (e.g. one that mimics NEURON) will have to apply the numbering requirement itself.

My preference would be to make this more flexible so that such validations are only implemented once, instead in an ad-hoc manner in each swc reader (which may or may not be a problem, depending on how many we end up having to write).

I will merge, because at the end of the day we are bike shedding over a poorly-specified widely-misused file format whose use we are unable to ban.